source/blog/2015-09-22-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,218 @@
+end
+```
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for it's
+arguments. These are higher order matchers (as opposed to more primitive
+matchers such as `be_predicate` or `be > x`) such as `include`, `start_with`
+and `match`. Let's take a look at the `match` matcher as a place to start.
+
+In RSpec 2.xx the match matcher was mostly used to match regexes, and its
+`matches?` definition looked thus:
+
+```ruby
+def matches?(expected, actual)
+  actual.match expected
+end
+```
+
+The new definition of this matcher's `matches?` method is thus:
+
+```ruby
+def matches?(expected, actual)
+  return true if values_match?(expected, actual)
+  return false unless can_safely_call_match?(expected, actual)
+  actual.match(expected)
+end
+```
+
+The key difference is that call to `values_match?` which leverages `===`
+(the core implementation of the new matching protocol in RSpec) to check if
+the "values match" and accepts matcher. The list of matchers which now use
+this approach are (as of RSpec expectations 3.3.0, not including mocks):
+
+* `change`
+* `contain_exactly`
+* `have_attributes`
+* `include`
+* `match`
+* `output`
+* `raise_error`
+* `start_with`
+* `end_with`
+* `throw_symbol`
+* `yield_control`
+* `yield_successive_args`
+* `yield_with_args`
+* `yield_with_no_args`
+
+Let's take a look at what `values_match?` does.
+
+`values_match?` is the RSpec match definition that allows matchers, and other objects, to compose.
+The implementation looks like this:
+
+```ruby
+def self.values_match?(expected, actual)
+  if Hash === actual
+    return hashes_match?(expected, actual) if Hash === expected
+  elsif Array === expected && Enumerable === actual && !(Struct === actual)
+    return arrays_match?(expected, actual.to_a)
+  end
+
+  return true if expected == actual
+
+  begin
+    expected === actual
+  rescue ArgumentError
+    # Some objects, like 0-arg lambdas on 1.9+, raise
+    # ArgumentError for `expected === actual`.
+    false
+  end
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
+
+The next thing the method does is to return `true` if `expected == actual`,
+note not `expected === actual` but `==` because `/foo/ === /foo/` returns
+`false` as does `MyClass === MyClass` -- so we have to check normal equality
+as well and not rely solely on `===`. We can demonstrate this is back in our
+IRB session:
+
+```ruby
+>> expect(/foo/).to match(/foo/)
+=> true
+```
+
+What we've done here is constructed a both a target and match matcher with the value /foo/.
+When the match matcher hits `values_match?` it sees that /foo/ is not an array or a hash, and so
+checks `expected == actual`, which in this case is `/foo/ == /foo/`, which is true.
```

Comment:
I found this paragraph (and the prior one) to be quite confusing, even though I authored the code in question.  Maybe we can come up with a simpler way to make this point?